backend/src/main/java/ai/verta/modeldb/lineage/LineageDAORdbImpl.java

```diff
@@ -30,205 +55,517 @@
-      validateExistence(
-          addLineage.getInputList(), addLineage.getOutputList(), isExistsPredicate, session);
+      List<LineageEntry> lineageEntries = new LinkedList<>(addLineage.getInputList());
+      lineageEntries.addAll(addLineage.getOutputList());
+      resourceExistsCheckConsumer.accept(session, lineageEntries);
       for (LineageEntry input : addLineage.getInputList()) {
-        for (LineageEntry output : addLineage.getOutputList()) {
-          addLineage(session, input, output);
-        }
+        addLineage(session, input, id, CONNECTION_TYPE_INPUT);
+      }
+      for (LineageEntry output : addLineage.getOutputList()) {
+        addLineage(session, output, id, CONNECTION_TYPE_OUTPUT);
       }
       session.getTransaction().commit();
+      return AddLineage.Response.newBuilder().setId(lineageElementEntity.getId()).build();
     }
-    return AddLineage.Response.newBuilder().setStatus(true).build();
   }
 
   @Override
-  public DeleteLineage.Response deleteLineage(DeleteLineage deleteLineage) throws ModelDBException {
+  public DeleteLineage.Response deleteLineage(DeleteLineage deleteLineage)
+      throws ModelDBException, InvalidProtocolBufferException {
     try (Session session = ModelDBHibernateUtil.getSessionFactory().openSession()) {
       session.beginTransaction();
       validate(deleteLineage.getInputList(), deleteLineage.getOutputList());
+      long id = deleteLineage.getId();
+      if (id == 0) {
+        throw new ModelDBException("Id not specified", Code.INVALID_ARGUMENT);
+      }
+      List<ConnectionEntity> connectionEntitiesById = getConnectionEntitiesById(session, id);
+      Map.Entry<
+              Map<LineageEntryContainer, ConnectionEntity>,
+              Map<LineageEntryContainer, ConnectionEntity>>
+          inputOutputs = getInputOutputs(session, connectionEntitiesById);
+      Map<LineageEntryContainer, ConnectionEntity> inputInDatabase = inputOutputs.getKey();
+      Map<LineageEntryContainer, ConnectionEntity> outputInDatabase = inputOutputs.getValue();
       for (LineageEntry input : deleteLineage.getInputList()) {
-        for (LineageEntry output : deleteLineage.getOutputList()) {
-          deleteLineage(session, input, output);
+        deleteLineage(session, input, inputInDatabase);
+      }
+      for (LineageEntry output : deleteLineage.getOutputList()) {
+        deleteLineage(session, output, outputInDatabase);
+      }
+      /* noDataProvided is a flag that there are no entries provided in the request.
+      That means that we should delete all entries with the specified id.
+      Also when there are no input or output in the database left we should delete
+      the remaining ones because they are not connected with anything. */
+      boolean noDataProvided =
+          deleteLineage.getInputCount() == 0 && deleteLineage.getOutputCount() == 0;
+      if (noDataProvided || inputInDatabase.isEmpty() || outputInDatabase.isEmpty()) {
```

Comment:
```suggestion
      boolean noDataProvided =
          deleteLineage.getInputCount() == 0 && deleteLineage.getOutputCount() == 0;
      if (noDataProvided || (inputInDatabase.isEmpty() && outputInDatabase.isEmpty())) {
```
The logic should be "there are no input **and** no output". It's fine to just have one in case the user is in the middle of editing.

Reply:
fixed